### PR TITLE
Convert the SRC component to use the module interface

### PR DESCRIPTION
**Pull request by gongjun-song**
Adopt module interface for src component.

Signed-off-by: Gongjun Song <gongjun.song@intel.com>

### CODE REVIEW

**Comment by gongjun-song** [comment]
The SRC module will have compatibility issues when implementing the module adapter on ipc3, so ipc3 retains the original implementation of comp_driver, and only implements the module adapter for ipc4.

Reasons for compatibility:
comp_driver executes the src_params function first, and then executes the dai_params function. The log is as follows ("ERROR gongjun: dai_verify_params()" is not a real error, just the log of the print position):
```
[    24094603.886316] (         129.270828) c0 src          2.9        src/audio/src/src.c:676  INFO src_params()
[    24094630.605065] (          26.718748) c0 src          2.9        src/audio/src/src.c:713  INFO src_params(), source_rate = 16000, sink_rate = 48000, format = 0
[    24094649.198815] (          18.593750) c0 src          2.9        src/audio/src/src.c:715  INFO src_params(), sourceb->channels = 2, sinkb->channels = 2, dev->frames = 48
[    24094794.042559] (         144.843750) c0 dai          2.14        src/ipc/ipc3/dai.c:110  INFO dai_data_config() dai type = 1 index = 0 dd 0x9e09b700
[    24094816.698808] (          22.656250) c0 dai          2.14    src/audio/dai-legacy.c:307  ERROR gongjun: dai_verify_params(): pcm rate parameter 48000, hardware rate 48000
```

If ipc3 is changed to module adapter, the dai_params function will be executed first, and then the src_params function will be executed (src_params is actually in the src_prepare function. "ERROR gongjun: dai_verify_params()" is not a real error, just the log of the print position).
```
[    15435984.647046] (          22.083332) c0 dai          2.14    src/audio/dai-legacy.c:306  ERROR gongjun: dai_verify_params(): pcm rate parameter 48000, hardware rate 48000
[    15436065.219959] (          80.572914) c0 dai          2.14    src/audio/dai-legacy.c:363  INFO dai_playback_params() dest_dev = 2 stream_id = 0 src_width = 4 dest_width = 4
[    15436085.428292] (          20.208332) c0 dai          2.14    src/audio/dai-legacy.c:369  INFO dai_playback_params() fifo 0x77010
[    15436124.178290] (          38.750000) c0 pipe         2.15  ....../pipeline-params.c:343  INFO pipe prepare
[    15436190.949121] (          66.770828) c0 src          2.9        src/audio/src/src.c:846  INFO src_prepare()
[    15436210.584537] (          19.635416) c0 src          2.9        src/audio/src/src.c:643  INFO src_params()
[    15436234.542869] (          23.958332) c0 src          2.9        src/audio/src/src.c:680  INFO src_params(), source_rate = 48000, sink_rate = 48000, format = 0
```

Explain why there is no error in the above module adapter log:
The log is to print the log when 48khz is selected, because if other rate is selected, an error will be reported at the position of dai_params, and the subsequent src_params cannot be executed, we unable to confirm the location of src_params when debugging . Like this: 
```
[    72535328.159372] (          22.291666) c0 dai          2.14    src/audio/dai-legacy.c:308  ERROR dai_verify_params(): pcm rate parameter 16000 does not match hardware rate 48000
[    72535384.826036] (          56.666664) c0 dai          2.14    src/audio/dai-legacy.c:483  ERROR dai_params(): pcm params verification failed.
[    72535426.544784] (          41.718750) c0 pipe         2.15  ....../pipeline-params.c:273  ERROR pipeline_params(): ret = -22, host->comp.id = 8
[    72535467.794783] (          41.250000) c0 ipc                  src/ipc/ipc3/handler.c:305  ERROR ipc: pipe 2 comp 8 params failed -22
[    72535508.732281] (          40.937500) c0 pipe         2.15  ......./pipeline-graph.c:400  INFO pipe reset
[    72535646.805192] (         138.072906) c0 src          2.9      src/audio/component.c:66   INFO comp_set_state(), state already set to 1
```

**Comment by gongjun-song** [comment]
SRC's Ipc4 module adapter depends on the PR: https://github.com/thesofproject/rimage/pull/116

File: src/audio/src/src.c [hunk 1]
```diff
@@ -54,36 +55,44 @@
...
+			 int *consumed,
+			 int *produced);
+#else
+	struct ipc4_config_src ipc_config;
+	void (*src_func)(struct processing_module *mod,
+			 struct input_stream_buffer *bsource,
+			 struct output_stream_buffer *bsink,
+			 int *n_read,
+			 int *n_written);
+#endif
```
Comment (lgirdwood) [inline src/audio/src/src.c:95]: These look very similar, are you sure we cant align them  - what did volume do ?

**Review by lgirdwood — COMMENTED** [review]
@singalsu pls review.

File: src/audio/src/src.c (hunk 1 above)
Comment (lgirdwood) [reply src/audio/src/src.c:95]: @gongjun-song will also need sparse support - @lyakh fyi.

File: src/audio/src/src.c [hunk 2]
```diff
@@ -624,12 +636,22 @@ static void src_free(struct comp_dev *dev)
...
 }
 
+static int src_stream_pcm_source_rate_check(struct ipc_config_src cfg,
+					    struct sof_ipc_stream_params *params)
+{
+	if (cfg.source_rate && params->rate != cfg.source_rate)
+		return -EINVAL;
+
+	return 0;
+}
```
Comment (lyakh) [inline src/audio/src/src.c:646]: could you keep this function where it was to reduce the diff?

File: src/audio/src/src.c [hunk 3]
```diff
@@ -624,12 +636,22 @@ static void src_free(struct comp_dev *dev)
...
+
 static int src_verify_params(struct comp_dev *dev,
 			     struct sof_ipc_stream_params *params)
 {
-	struct comp_data *cd = comp_get_drvdata(dev);
 	int ret;
 
+	struct comp_data *cd = comp_get_drvdata(dev);
```
Comment (lyakh) [inline src/audio/src/src.c:653]: same here - better avoid unnecessary changes

File: src/audio/src/src.c [hunk 4]
```diff
@@ -815,52 +832,6 @@ static int src_trigger(struct comp_dev *dev, int cmd)
 	return comp_set_state(dev, cmd);
 }
 
-static int src_get_copy_limits(struct comp_data *cd,
-			       const struct comp_buffer __sparse_cache *source,
-			       const struct comp_buffer __sparse_cache *sink)
-{
```
Comment (lyakh) [inline src/audio/src/src.c:821]: please keep this function where it was if possible

File: src/audio/src/src.c [hunk 5]
```diff
@@ -323,6 +335,49 @@ int src_polyphase_init(struct polyphase_src *src, struct src_param *p,
 	return n_stages;
 }
 
+static int src_check_buffer_sizes(struct comp_dev *dev, struct comp_data *cd,
+				  struct audio_stream __sparse_cache *source_stream,
+				  struct audio_stream __sparse_cache *sink_stream)
```
Comment (lyakh) [inline src/audio/src/src.c:340]: please keep this function where it was if possible

File: src/audio/src/src.c [hunk 6]
```diff
@@ -1082,3 +1008,623 @@ UT_STATIC void sys_comp_src_init(void)
...
+{
+	*n_read = 0;
+	*n_written = 0;
+	src_update_buffer_position(bsource, bsink, n_read, n_written);
+}
+
+/* Normal 2 stage SRC */
+static void src_2s(struct processing_module *mod, struct input_stream_buffer *bsource,
+		   struct output_stream_buffer *bsink, int *n_read, int *n_written)
+{
```
Comment (lyakh) [inline src/audio/src/src.c:1036]: can we try to avoid duplicating these functions and create a common function instead and call it from two wrappers - for IPC3 and IPC4 / module adapter?

**Review by lyakh — CHANGES_REQUESTED** [review]
It looks like it should be possible to reduce the size of this patch a lot, this would make reviewing it easier and reduce the probability of errors

**Comment by marc-hb** [comment]
> It looks like it should be possible to reduce the size of this patch a lot, this would make reviewing it easier and reduce the probability of errors

Not just a review issue:

https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes

File: src/audio/src/src.c (hunk 1 above)
Comment (gongjun-song) [reply src/audio/src/src.c]: thanks, updated

File: src/audio/src/src.c [hunk 7]
```diff
@@ -54,36 +55,44 @@
...
-DECLARE_SOF_RT_UUID("src", src_uuid, 0xe61bb28d, 0x149a, 0x4c1f,
-		    0xb7, 0x09, 0x46, 0x82, 0x3e, 0xf5, 0xf5, 0xae);
-#elif CONFIG_IPC_MAJOR_3
+#if CONFIG_IPC_MAJOR_3
+static const struct comp_driver comp_src;
+
 /* c1c5326d-8390-46b4-aa47-95c3beca6550 */
 DECLARE_SOF_RT_UUID("src", src_uuid, 0xc1c5326d, 0x8390, 0x46b4,
 		    0xaa, 0x47, 0x95, 0xc3, 0xbe, 0xca, 0x65, 0x50);
 #else
```
[the remark this answers is not in the document]
Comment (gongjun-song) [reply src/audio/src/src.c]: thanks, updated

File: src/audio/src/src.c [hunk 8]
```diff
@@ -54,36 +55,44 @@
...
-static const struct comp_driver comp_src;
-
 LOG_MODULE_REGISTER(src, CONFIG_SOF_LOG_LEVEL);
 
-#if CONFIG_IPC_MAJOR_4
-/* e61bb28d-149a-4c1f-b709-46823ef5f5a3 */
-DECLARE_SOF_RT_UUID("src", src_uuid, 0xe61bb28d, 0x149a, 0x4c1f,
-		    0xb7, 0x09, 0x46, 0x82, 0x3e, 0xf5, 0xf5, 0xae);
-#elif CONFIG_IPC_MAJOR_3
+#if CONFIG_IPC_MAJOR_3
```
[the remark this answers is not in the document]
Comment (gongjun-song) [reply src/audio/src/src.c]: updated

File: src/audio/Kconfig [hunk 9]
```diff
@@ -62,6 +62,7 @@ config COMP_BLOB
 
 config COMP_SRC
 	bool "SRC component"
+	depends on COMP_MODULE_ADAPTER
```
[the remark this answers is not in the document]
Comment (gongjun-song) [reply src/audio/Kconfig]: yes, only ipc4 support module adapter, I removed some setup code from Kconfig and CMakeLists.txt. If this is not a correct way, please give me some advice.

File: src/audio/src/src.c (hunk 3 above)
Comment (gongjun-song) [reply src/audio/src/src.c]: updated

File: src/audio/src/src.c (hunk 4 above)
Comment (gongjun-song) [reply src/audio/src/src.c]: updated

File: src/audio/src/src.c (hunk 5 above)
Comment (gongjun-song) [reply src/audio/src/src.c]: Because the location of the code referring to the "src_check_buffer_sizes" function has changed, the location needs to be changed

File: src/audio/src/src.c (hunk 6 above)
Comment (gongjun-song) [reply src/audio/src/src.c:942]: updated

File: src/audio/src/src.c (hunk 2 above)
Comment (gongjun-song) [reply src/audio/src/src.c:584]: > 

Thanks‘s for suggestion, I have changed code.

Comment (gongjun-song) [reply src/audio/src/src.c:584]: updated

**Comment by gongjun-song** [comment]
> > It looks like it should be possible to reduce the size of this patch a lot, this would make reviewing it easier and reduce the probability of errors
> 
> Not just a review issue:
> 
> https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes

Thanks‘s for suggestion, I have changed code.

**Comment by marc-hb** [comment]
The compilation failures in https://github.com/thesofproject/sof/actions/runs/3523675882/jobs/5908116892 seem relevant

**Comment by gongjun-song** [comment]
> The compilation failures in https://github.com/thesofproject/sof/actions/runs/3523675882/jobs/5908116892 seem relevant

Thanks, have fixed.

File: src/audio/src/src.c [hunk 10]
```diff
@@ -545,6 +583,409 @@ static void src_set_sink_params(struct comp_dev *dev, struct comp_buffer __spars
...
+	struct audio_stream __sparse_cache *source = input_buffers->data;
+	struct audio_stream __sparse_cache *sink = output_buffers->data;
+
+	input_buffers->consumed += audio_stream_frame_bytes(source) * (*n_read);
+	output_buffers->size += audio_stream_frame_bytes(sink) * (*n_written);
+}
+
+static int src_verify_params(struct comp_dev *dev,
+			     struct sof_ipc_stream_params *params)
+{
```
Comment (lyakh) [inline src/audio/src/src.c:599]: you're still duplicating code. What I meant was:
```
@@ -624,10 +624,10 @@ static void src_free(struct comp_dev *dev)
 	rfree(dev);
 }
 
-static int src_verify_params(struct comp_dev *dev,
-			     struct sof_ipc_stream_params *params)
+static int src_verify_params_intern(struct comp_dev *dev,
+				    struct sof_ipc_stream_params *params,
+				    struct comp_data *cd)
 {
-	struct comp_data *cd = comp_get_drvdata(dev);
 	int ret;
 
 	comp_dbg(dev, "src_verify_params()");
@@ -661,6 +661,14 @@ static int src_verify_params(struct comp_dev *dev,
 	return ret;
 }
 
+static int src_verify_params(struct comp_dev *dev,
+			     struct sof_ipc_stream_params *params)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	return src_verify_params_intern(dev, params, cd);
+}
+
 /* set component audio stream parameters */
 static int src_params(struct comp_dev *dev,
 		      struct sof_ipc_stream_params *params)
```
and then for the module adapter version:
```
+static int src_verify_params(struct comp_dev *dev,
+			     struct sof_ipc_stream_params *params)
+{
+	struct processing_module *mod = comp_get_drvdata(dev);
+	struct comp_data *cd = module_get_private_data(mod);
+
+	return src_verify_params_intern(dev, params, cd);
+}
```
This should let you make the diff quite small and also avoid adding duplicate code to the file.

**Comment by singalsu** [comment]
SRC is very tricky, how has this been tested? The IPC3 SRC has been tested in testbench with tools/test/audio/src_test.m. But there's no IPC4 support in testbench.

File: src/audio/src/src.c (hunk 10 above)
Comment (gongjun-song) [reply src/audio/src/src.c:595]: thanks for your advice. For the convenience of review, I split the previous commit into several.

**Comment by gongjun-song** [comment]
> SRC is very tricky, how has this been tested? The IPC3 SRC has been tested in testbench with tools/test/audio/src_test.m. But there's no IPC4 support in testbench.

Due to commit 1dffde13ba4b181d483ee54, the testbench execution failed(program hangs), so I cannot use testbench to test the src of ipc3 (it seems like a bug, neither the version I wrote nor the default version of sof can work).

But whether it is ipc3 or ipc4, I use different rates to play audio, the sound is played normally, and there is no noise.
Test command:
$: aplay -t wav -r 16000 -Dplughw:0,0 ./audio_48k_10s_s16.wav
$: aplay -t wav -r 44100 -Dplughw:0,0 ./audio_48k_10s_s44.wav
$: aplay -t wav -r 48000 -Dplughw:0,0 ./audio_48k_10s_s48.wav

**Comment by lgirdwood** [comment]
@ranj063 can you take a look too, thanks :) Want to get all modules done for v2.5

**Comment by marc-hb** [comment]
> Due to commit https://github.com/thesofproject/sof/commit/1dffde13ba4b181d483ee54f0a0df2b4cee5c5db, the testbench execution failed(program hangs), so I cannot use testbench to test the src of ipc3 (it seems like a bug, neither the version I wrote nor the default version of sof can work).


https://github.com/thesofproject/sof/actions/runs/3565026602/jobs/5989687227 is green. Should it be red?

**Comment by gongjun-song** [comment]
> > Due to commit [1dffde1](https://github.com/thesofproject/sof/commit/1dffde13ba4b181d483ee54f0a0df2b4cee5c5db), the testbench execution failed(program hangs), so I cannot use testbench to test the src of ipc3 (it seems like a bug, neither the version I wrote nor the default version of sof can work).
> 
> https://github.com/thesofproject/sof/actions/runs/3565026602/jobs/5989687227 is green. Should it be red?

There is no problem with the test results of CI, I am not using the same script as the CI test.
CI uses ./scripts/host-testbench.sh, and I use tools/test/audio/src_test.sh for testing.

The script src_test.sh also test fails with default SOF on my machine, which is not caused by this PR.
In addition, the SRC ipc3 in this PR still maintains the original comp_driver processing method.

**Comment by singalsu** [comment]
> > SRC is very tricky, how has this been tested? The IPC3 SRC has been tested in testbench with tools/test/audio/src_test.m. But there's no IPC4 support in testbench.
> 
> Due to commit [1dffde1](https://github.com/thesofproject/sof/commit/1dffde13ba4b181d483ee54f0a0df2b4cee5c5db), the testbench execution failed(program hangs), so I cannot use testbench to test the src of ipc3 (it seems like a bug, neither the version I wrote nor the default version of sof can work).
> 
> But whether it is ipc3 or ipc4, I use different rates to play audio, the sound is played normally, and there is no noise. Test command: $: aplay -t wav -r 16000 -Dplughw:0,0 ./audio_48k_10s_s16.wav $: aplay -t wav -r 44100 -Dplughw:0,0 ./audio_48k_10s_s44.wav $: aplay -t wav -r 48000 -Dplughw:0,0 ./audio_48k_10s_s48.wav

The subjective listening can easily miss a small issue like -60 dB THD+N vs. -90 dB THD+N. The src_test.m modification to operate on a real device with nocodec SSP loopback would be a reliable quality test. Use aplay and arecord instead of testbench launch in the script. Could such enhance be taken to plans? Also enhancing testbench for IPC4 would be essential since we are dropping IPC3 from future releases.

**Comment by singalsu** [comment]
@gongjun-song I think I can fix the topology that src_test.m uses, increasing source and sink buffer should help the freezing 11.025 kHz rates to work. I'll try it today. Also testbench needs some max copy() count check and error to prevent freezes.

**Review by lyakh — CHANGES_REQUESTED** [review]
Thanks for splitting your first version of this PR. Indeed, separating moving functions around to new locations into separate commits. I still think it could be done with less moving around. Your reason for moving functions is to keep only one major `#if CONFIG_IPC_MAJOR_4` switch in the code, but you already have two of them. So, if you wanted to have one - yes, you have to move code around. But if you're fine with having multiple of them - you should be able to keep most of the current code in place. But that's less important, I guess. But buffer acquisition / releasing must be fixed

**Comment by singalsu** [comment]
@gongjun-song My (draft) fix for SRC freeze is https://github.com/thesofproject/sof/pull/6688

**Comment by singalsu** [comment]
> Whenever someone says "I'm testing something different from CI" then we have a serious test gap almost every time. It's a big red flag; absolutely not OK. Every product is only as good as the tests it's passing.
> 
> @singalsu can you please add `src_test.sh` to PR testing? Unless your #6668 already does something equivalent?

Yep, would adding it to scripts/host-testbench.sh be OK? It would add about 1-2 minute of testing time as quick test.

**Comment by marc-hb** [comment]
> Yep, would adding it to scripts/host-testbench.sh be OK? 

It depends: is `src_test.sh` related / similar to `host-testbench.sh`? If not then it's better to keep them separate and append two lines at the bottom of `.github/workflows/testbench.yml`

By the way I had a look at `src_test.sh` and the little it does looks silly: why can't `octave` print PASS/FAIL itself?

> It would add about 1-2 minute of testing time as quick test.

This is a question only for the main developers = you: for how long are you ready to wait for test results?

File: src/audio/src/src.c [hunk 11]
```diff
@@ -528,7 +529,8 @@ static void src_set_params(struct comp_dev *dev, struct sof_ipc_stream_params *p
 	params->buffer.size = cd->ipc_config.base.ibs;
 
 	sinkb = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
-	sinkb->stream.rate = cd->ipc_config.sink_rate;
+	sink_c = buffer_acquire(sinkb);
+	sink_c->stream.rate = cd->ipc_config.sink_rate;
```
[the remark this answers is not in the document]
Comment (gongjun-song) [reply src/audio/src/src.c:528]: thanks, updated

File: src/audio/src/src.c [hunk 12]
```diff
@@ -880,6 +888,162 @@ static int src_prepare_general(struct comp_dev *dev, struct comp_data *cd,
...
+	sourceb = list_first_item(&dev->bsource_list, struct comp_buffer, sink_list);
+	sinkb = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
+
+	source_c = buffer_acquire(sourceb);
+	sink_c = buffer_acquire(sinkb);
+
+	/* set align requirements */
+	src_set_alignment(&source_c->stream, &sink_c->stream);
+
+	ret = src_prepare_general(dev, cd, source_c, sink_c);
```
[the remark this answers is not in the document]
Comment (gongjun-song) [reply src/audio/src/src.c]: updated

File: src/audio/src/src.c [hunk 13]
```diff
@@ -880,6 +888,162 @@ static int src_prepare_general(struct comp_dev *dev, struct comp_data *cd,
...
+	ret = src_params_general(dev, cd, params);
+	if (ret < 0)
+		return ret;
+
+	/* SRC component will only ever have 1 source and 1 sink buffer */
+	sourceb = list_first_item(&dev->bsource_list, struct comp_buffer, sink_list);
+	sinkb = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
+
+	source_c = buffer_acquire(sourceb);
+	sink_c = buffer_acquire(sinkb);
```
[the remark this answers is not in the document]
Comment (gongjun-song) [reply src/audio/src/src.c:948]: src_prepare function's acquired buffers will be released in src_prepare_general function. Don't need to update.

File: src/audio/src/src.c [hunk 14]
```diff
@@ -880,6 +888,162 @@ static int src_prepare_general(struct comp_dev *dev, struct comp_data *cd,
...
+	memcpy_s(&cd->ipc_config, sizeof(cd->ipc_config), cfg->init_data, sizeof(cd->ipc_config));
+
+	cd->delay_lines = NULL;
+	cd->src_func = src_fallback;
+	cd->polyphase_func = NULL;
+	src_polyphase_reset(&cd->src);
+
+	md->private = cd;
+
+	mod->simple_copy = true;
```
[the remark this answers is not in the document]
Comment (gongjun-song) [reply src/audio/src/src.c:923]: src_init and comp_new have different return value and spec value, so there is less common code. I recommend keeping the current code.

**Comment by gongjun-song** [comment]
> @gongjun-song My (draft) fix for SRC freeze is #6688

Thank you, based on the https://github.com/thesofproject/sof/pull/6688, there is no problem running src_test.sh with SRC ipc3 of this PR, and all cases have passed. 
For IPC4, @andrula-song will help write testbench for SRC in 2023-Q1.

@singalsu, in addition to the above testbench issue, could you please review the code to see if there are any other issues? thanks!

**Comment by lgirdwood** [comment]
> > @gongjun-song My (draft) fix for SRC freeze is #6688
> 
> Thank you, based on the #6688, there is no problem running src_test.sh with SRC ipc3 of this PR, and all cases have passed. For IPC4, @andrula-song will help write testbench for SRC in 2023-Q1.
> 
> @singalsu, in addition to the above testbench issue, could you please review the code to see if there are any other issues? thanks!

#6688 now merged - @singalsu @ranj063 pls review.

File: src/audio/src/src.c [hunk 15]
```diff
@@ -870,6 +823,126 @@ static int src_params(struct comp_dev *dev,
...
+		ret = -EINVAL;
+		goto out;
+	}
+
+out:
+	if (ret < 0)
+		comp_set_state(dev, COMP_TRIGGER_RESET);
+
+	buffer_release(sink);
+	buffer_release(source);
```
Comment (lyakh) [inline src/audio/src/src.c:888]: nnno, let's not do this. Formally this is correct and reduces code duplication, but I don't think it's a good coding practice. Usually you want to acquire and release or lock and unlock objects at the same level, within the same function. It's too error prone the way you do it here. Let's keep these two `buffer_release()` lines in the caller functions, or maybe you could move `buffer_acquire()` inside this function too. But let's not split it between the caller and here.

File: src/audio/src/src.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (lyakh) [reply src/audio/src/src.c:949]: you could either add
```
	/* SRC component will only ever have 1 source and 1 sink buffer */
	sourceb = list_first_item(&dev->bsource_list, struct comp_buffer, sink_list);
	sinkb = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);

	source_c = buffer_acquire(sourceb);
	sink_c = buffer_acquire(sinkb);

#if CONFIG_IPC_MAJOR_4
	/* set align requirements */
	src_set_alignment(&source_c->stream, &sink_c->stream);
#endif
```
to `src_prepare_general()` or move the two `buffer_release()` calls from it to the caller, but let's keep acquire / release symmetric.

File: src/audio/src/src.c [hunk 16]
```diff
@@ -891,6 +905,152 @@ static int src_prepare_general(struct comp_dev *dev, struct comp_data *cd)
...
+}
+
+static int src_free(struct processing_module *mod)
+{
+	struct comp_data *cd = module_get_private_data(mod);
+
+	comp_info(mod->dev, "src_free()");
+
+	/* Free dynamically reserved buffers for SRC algorithm */
+	if (cd->delay_lines)
```
Comment (lyakh) [inline src/audio/src/src.c:1034]: `if` is not needed, `rfree(NULL)` is allowed

**Review by lyakh — APPROVED** [review]
LGTM now, just a couple of minor comments

File: src/audio/src/src.c (hunk 16 above)
Comment (gongjun-song) [reply src/audio/src/src.c]: Thanks, I also added a new commit fixing the previous code.

File: zephyr/wrapper.c [hunk 17]
```diff
@@ -287,7 +289,11 @@ int task_main_start(struct sof *sof)
 		sys_comp_dai_init();
 
 	if (IS_ENABLED(CONFIG_COMP_SRC))
+#if CONFIG_IPC_MAJOR_3
 		sys_comp_src_init();
+#else
+		sys_comp_module_src_interface_init();
+#endif
```
[the remark this answers is not in the document]
Comment (gongjun-song) [reply zephyr/wrapper.c:296]: updated

[the remark this answers is not in the document]
Comment (lyakh) [reply zephyr/wrapper.c:296]: thanks!

**Review by lgirdwood — APPROVED** [review]
